Unify group iso functions for changing group type

lgoettgens opened the pull request:
As proposed by @fingolfin in https://github.com/oscar-system/Oscar.jl/pull/4614#discussion_r1962456148.

lgoettgens committed:
Unify `group_type(::Group)` methods

lgoettgens committed:
Import `AbstractAlgebra.Group`

lgoettgens committed:
Unify `(::Group)(::Group)` methods

lgoettgens commented on src/Groups/homomorphisms.jl:
```diff
@@ -1191,23 +1191,22 @@ end
-end
-
-function (::Type{T})(G::FinGenAbGroup) where T <: GAPGroup
+function (::Type{T})(G::Group) where T <: Group
```
This is, unfortunately, type piracy as we neither own `Group` nor `Type{Group}`, and thus this also defined methods for e.g. `AbstractAlgebra.SymmetricGroup(::AbstractAlgebra.SymmetricGroup)`.
We cannot use a large Union here since `WeylGroup` is only defined much later, and changing the order just leads to other problems.
One possible solution would be to introduce an abstract type `OscarGroup <: Group` and then say `GAPGroup <: OscarGroup` and `WeylGroup <: OscarGroup` and at this place here require that at least one of the args is a `OscarGroup`. (this would need 3 methods: `Group(::OscarGroup)`, `OscarGroup(::Group)` and for disambiguation `OscarGroup(::OscarGroup)`.

And then we have the problem that `FinGenAbGrp <: Group` does not hold, so we would need to add more dispatches for that again...

All of this only considers the third commit, the other two should be fine to change.

Any idea what to do @fingolfin ?

fingolfin replied:
How about putting this into AbstractAlgebra? Together with a generic `isomorphism` implementation which basically only implements something like `isomorphism(::Type{T}, G::T) = identity_hom(G)` (pseudo code)

`FinGenAbGrp` is an oddball because it is additive; as such I could live with having to handle it separately...

lgoettgens replied:
I'll try that next week (other things to do for today)

fingolfin commented on src/Groups/homomorphisms.jl:
```diff
@@ -1191,23 +1191,22 @@ end
 
 Return a group of the requested type that is isomorphic to `G`.
+An exception is thrown if no such group exists.
+A MethodError is thrown if this particular pair of types is not implemented (yet).
```
```suggestion
A `MethodError` is thrown if this particular pair of types is not implemented (yet).
```
Makes one wonder in the line above what error is thrown in that case... :-)

ThomasBreuer replied:
If one tries `permutation_group(free_group(1))` then one gets an `ArgumentError`.
I am not sure whether we can guarantee this in general, because it might happen that we run into a GAP error instead of getting a `fail` result from GAP.
Should we say this in the documentation?
And should we better use `try`/`catch` in such situations?

lgoettgens replied:
I deliberately didn't specify the error type in that case as I don't know if all current functions throw consistent error types, and if we want to specify the error type at all.

ThomasBreuer reviewed (commented):
Concerning the type piracy problem:

In several places of the documentation, we try to give a general definition for `Group` type arguments, but separate methods are needed for `GAPGroup`, `FinGenAbGroup`, `MultTableGroup` and so on.

This is cheating in the sense that the implementation does not match the documentation (as the example of `AbstractAlgebra.Generic.SymmetricGroup` shows), and some people got already confused about the fact that the signatures in the docstrings do not fit to the signatures of the corresponding methods.

On the other hand, I like the idea to document a general concept (a function) only once, and then various methods (have to) implement what this documentation promises.

This is how it is done in GAP: Documentation is provided mainly for functions (called operations in GAP), only in special cases also individual methods have their own documentation. A disadvantage of this approach is that the documentation then often promises more than the implementation provides, because methods are missing or because in general there is no guarantee that the promised result can be computed at all.

Thus the conceptual question is whether we want to have documentation in the sense of general "declarations" (like in GAP) what a function returns, or we want to document which methods are really available.

The technical question is how to create a useful type hierarchy. If we want to avoid a type union, and introduce one `OscarGroup` type on the top then one of the problems is that we want `FinGenAbGroup` under `OscarGroup`, but `FinGenAbGroup` is a subtype of `AbstractAlgebra.AdditiveGroup`.

lgoettgens commented:
> The technical question is how to create a useful type hierarchy. If we want to avoid a type union, and introduce one `OscarGroup` type on the top then one of the problems is that we want `FinGenAbGroup` under `OscarGroup`, but `FinGenAbGroup` is a subtype of `AbstractAlgebra.AdditiveGroup`.

No, for the purpose I was talking about here, we don't want to include `FinGenAbGroup`, as that is not *owned* by Oscar (in the type piracy sense). If it would be included, everything like `(::Type{OscarGroup})(::OscarGroup)` would be type piracy.

Thinking about it again, I think we need two abstract types: `GroupDefinedInOscar <: Group` would be the abstract supertype of `GAPGroup` and `WeylGroup`. And then we have something like `const OscarGroup = Union{GroupDefinedInOscar, FinGenAbGroup, MultTableGroup}`. Signatures would then always mention `OscarGroup`. But we would be able to define these conversions here without type piracy by providing `(::Type{OscarGroup})(::GroupDefinedInOscar)` and `(::Type{GroupDefinedInOscar})(::OscarGroup)` (and `(::Type{GroupDefinedInOscar})(::GroupDefinedInOscar)` for disambiguation).

We could even name `GroupDefinedInOscar <: AbstractAlgebra.Group` just `Group <: AbstractAlgebra.Group` as the supertype in AA is not exported there, and not imported into Oscar (until this preliminary PR, but that could be changed).

fingolfin reviewed (commented):
Overall I like it, thank you

ThomasBreuer commented:
> Thinking about it again, I think we need two abstract types: `GroupDefinedInOscar <: Group` would be the abstract supertype of `GAPGroup` and `WeylGroup`. And then we have something like `const OscarGroup = Union{GroupDefinedInOscar, FinGenAbGroup, MultTableGroup}`.[...]

As far as I understand this, then any new type `T` of groups that gets introduced in some package depending on Oscar can be defined as a subtype of `GroupDefinedInOscar`, which means that the documented functionality for `OscarGroup` shall hold for groups of type `T`, and generic methods for `OscarGroup` will automatically be applicable.

lgoettgens committed:
Adapt to new stub in AA

lgoettgens committed:
More consistent `on_gens` usage

lgoettgens commented:
I now moved some function stubs to AA (see https://github.com/Nemocas/AbstractAlgebra.jl/pull/2036), moved everything that was just dealing with FinGenAbGroups (and no Oscar-defined groups) to https://github.com/thofma/Hecke.jl/pull/1800.
This makes it possible to do everything we want to do in Oscar without adding any of the previously mentioned abstract types, and without type piracy.

CI is expected to fail here; the relevant CI for the moment is the downstream CI in https://github.com/Nemocas/AbstractAlgebra.jl/pull/2036.

Another change here is that I add the kwarg `on_gens` to all `isomorphism(::Type, ::Union{Group,FinGenAbGroup})` specializations to allow for more generic usage. All methods that previously didn't allow this kwarg will throw when a non-default value is passed.

lgoettgens commented:
According to the downstream CI in https://github.com/Nemocas/AbstractAlgebra.jl/pull/2036, everything works as I expected. So this is ready for review now (but remember that the review needs to involve all three PRs in different repositiories).

lgoettgens committed:
Remove another redundant method

ThomasBreuer reviewed (approved):
Looks o.k.

Just for curiosity:
Why is it useful to support the `on_gens` argument in those `isomorphism` methods where it was not supported, and then to throw an exception if one sets it to `true` in an `isomorphism` call (or to ignore the value)?

thofma commented:
Can you concisely summarize what the changes are/what the goal is? Would be ideal if it did not require to interpret https://github.com/oscar-system/Oscar.jl/pull/4614#discussion_r1962456148.

lgoettgens commented:
After these three PRs are merged, the process for adding a new such iso is the following (example taken from the Weyl groups that gets adapted in this PR to be like I describe here).

```julia
function isomorphism(T::Type{FPGroup}, W::WeylGroup; on_gens::Bool=false)
  # If iso only possible for `on_gens=false`, throw if `on_gens == true`.
  # If gens get always mapped to gens, include the next line.
  on_gens = true # we ignore the on_gens flag, the iso will *always* map gens onto gens
  isos =
    get_attribute!(Dict{Tuple{Type,Bool},Any}, W, :isomorphisms)::Dict{Tuple{Type,Bool},Any}
  return get!(isos, (T, on_gens)) do
    # do the work
    return <the iso>
  end::<type of the iso>
end
```

The generic code in the PRs then also allow `FPGroup(::WeylGroup)` and `fp_group(::WeylGroup)` by redirecting to `codomain(isomorphism(FPGroup, ::WeylGroup))`.

Additionally, there is automatically `isomorphism(WeylGroup, ::WeylGroup)` which constructs an identity home.

The changes in Hecke additionally allow either of the two group types involved to be a `FinGenAbGroup`. (And there is some moving things around to make all of these function type piracy free)

fingolfin commented:
This has a conflict that needs to be resolved

fingolfin reviewed (approved):
Looks good to me.

lgoettgens committed:
Put Hecke changes into Oscar for the time being

lgoettgens committed:
Bump AA compat

lgoettgens commented:
Since https://github.com/thofma/Hecke.jl/pull/1800 introduces some incompatibilities with Oscar 1.3.1 (or rather any Oscar without this PR here), I instead included the Hecke changes in `misc/MoveToHecke.jl`, from where we can move it to Hecke with a future breaking Hecke release.

Once CI is green, this is good to go from my side.

lgoettgens committed:
Fix typo
